src/protocol/sort/generate_permutation_opt.rs

```diff
@@ -109,39 +114,157 @@ where
-                    .randoms_for_shuffle
-                    .0
-                    .as_slice(),
-                revealed_and_random_permutations
-                    .randoms_for_shuffle
-                    .1
-                    .as_slice(),
-            ),
-            &revealed_and_random_permutations.revealed,
+            (randoms_for_shuffle0, randoms_for_shuffle1),
+            revealed,
             next_few_bits_permutation,
         )
         .await?;
     }
     Ok(composed_less_significant_bits_permutation)
 }
 
+#[allow(dead_code)]
+/// Returns a sort permutation in a malicious context.
+/// This runs sort in a malicious context. The caller is responsible to validate the accumulater contents and downgrade context to Semi-honest before calling this function
+/// The function takes care of upgrading and validating while the sort protocol runs.
+/// It then returns a semi honest context with output in Replicated format. The caller should then upgrade the output and context before moving forward
+///
+/// Steps
+/// 1. [Malicious Special] Upgrade the context from semihonest to malicious and get a validator
+/// 2. [Malicious Special] Upgrade 0..`num_multi_bits` sort bit keys
+/// 3. Compute bit permutation that sorts 0..`num_multi_bits` bit
+///
+/// For `num_multi_bits` to N-1th bit of input share
+/// 1. i. Shuffle the i-1th composition
+///   ii. [Malicious Special] Validate the accumulator contents
+///  iii. [Malicious Special] Malicious reveal
+///   iv. [Malicious Special] Downgrade context to semihonest
+/// 2. i. [Malicious Special] Upgrade ith sort bit keys
+///   ii. Sort i..i+`num_multi_bits` bits based on i-1th bits by applying i-1th composition on i..i+`num_multi_bits` bits
+/// 3. Compute bit permutation that sorts i..i+`num_multi_bits` bits
+/// 4. Compute ith composition by composing i-1th composition on ith permutation
+/// In the end, following is returned
+///    i. n-1th composition: This is the permutation which sorts the inputs
+///   ii. Validator which can be used to validate the leftover items in the accumulator
+///
+/// # Panics
+/// If sort keys dont have num of bits same as `num_bits`
+/// # Errors
+pub async fn malicious_generate_permutation_opt<'a, F>(
+    sh_ctx: SemiHonestContext<'a, F>,
+    sort_keys: &[Vec<Replicated<F>>],
+    num_bits: u32,
+    num_multi_bits: u32,
+) -> Result<(MaliciousValidator<'a, F>, Vec<MaliciousReplicated<F>>), Error>
+where
+    F: Field,
+{
+    let mut malicious_validator = MaliciousValidator::new(sh_ctx.narrow(&MaliciousUpgradeContext));
+    let mut m_ctx = malicious_validator.context();
+    let m_ctx_0 = m_ctx.narrow(&Sort(0));
+    assert_eq!(sort_keys.len(), num_bits as usize);
+
+    let last_bit_num = std::cmp::min(num_multi_bits, num_bits);
+
+    let upgraded_sort_keys = try_join_all(
+        (0..last_bit_num)
+            .zip(repeat(m_ctx.clone()))
+            .map(|(i, m_ctx)| async move { m_ctx.upgrade(sort_keys[i as usize].clone()).await }),
```

Comment:
@andyleiserson this still will fail since we are using the same context across bits for the same record with the error "Generated randomness for index '0' twice using the same key 'protocol/run-0/sort/malicious_upgrade_context/upgrade_input' ". Any ideas how to resolve this without narrowing context?

Reply:
My thinking (based on an earlier revision of this PR) was to [add an upgrade impl for `Vec<Vec<Replicated>>`](https://github.com/private-attribution/ipa/commit/b80bfe43f43aa0d4f9af4b6a57bbf9548a5b387c) and then [use that to upgrade everything at once rather than doing multiple upgrades](https://github.com/private-attribution/ipa/commit/173ac7b57416deb917cfe84c2007ba8a6aa6d74f).

Does that make sense? I'm not sure if something changed since I looked at it or if there's some other reason I missed that that won't work.

Reply:
The code was earlier written to just work on one vector at a time and so I am doing some refactoring to let the sort code take in vec<vec<vec<replicated>>> so that each vec<vec<replicated>> can be processed in one iteration